Java/app/src/main/java/com/taboola/sdk3example/sdk_native/WidgetFragment.java

```diff
@@ -0,0 +1,115 @@
+            nativeUnit.fetchRecommendations(requestData, new TBLRecommendationRequestCallback() {
+                @Override
+                public void onRecommendationsFetched(TBLRecommendationsResponse tblRecommendationsResponse) {
+                    println("Taboola | fetchInitialContent | onRecommendationsFetched");
+                    // Add response content to layout
+                    displayContent(contentLayout, tblRecommendationsResponse);
+                }
+
+                @Override
+                public void onRecommendationsFailed(Throwable throwable) {
+                    println(String.format("Taboola | onRecommendationsFailed: %s", throwable.getLocalizedMessage()));
+                }
+            });
+        }
+
+        return root;
+    }
+
+    /**
+     * Define a Page that represents this screen, get a Unit from it, add it to screen and fetch its content
+     * Notice: A Unit of unlimited items, called "Feed" in Taboola, can be set in TBL_PLACEMENT_TYPE.PAGE_BOTTOM only.
+     */
+    private TBLNativeUnit getTaboolaUnit() {
+        // Define a page to control all Unit placements on this screen
+        TBLNativePage nativePage = Taboola.getNativePage("text", "https://blog.taboola.com");
+
+        return nativePage.build("list_item", new TBLNativeListener() {
+            @Override
+            public boolean onItemClick(String placementName, String itemId, String clickUrl, boolean isOrganic, String customData) {
+                println(String.format("Taboola | onItemClick | isOrganic = %s", isOrganic));
+                return super.onItemClick(placementName, itemId, clickUrl, isOrganic, customData);
+            }
+        });
+    }
+
+    /**
+     * In Native integrations fetch request returns multiple Android Views.
+     * It is up to you to lay them out.
+     */
+    private void displayContent(LinearLayout contentLayout, TBLRecommendationsResponse tblRecommendationsResponse) {
+        if (tblRecommendationsResponse == null || tblRecommendationsResponse.getPlacementsMap() == null) {
+            println("Error: No recommendations returned from server.");
+            return;
+        }
+        TBLPlacement placement = tblRecommendationsResponse.getPlacementsMap().values().iterator().next();
+
+        if (placement != null && placement.getItems().size() > 0) {
+
+            // We asked for 4 item so we go over all items returned from Taboola
+            // NOTE: In this example, we search for the items relevant for the only placement in this page
+
+            for (TBLRecommendationItem item : placement.getItems()) {
+                Context context = contentLayout.getContext();
+                ViewGroup.LayoutParams layoutParams = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
+                TBLImageView thumbnailView = item.getThumbnailView(context);
+                TBLTextView titleView = item.getTitleView(context);
```

Comment:
Maybe put all the little parts of the Native UI in a method, explaining this unique behavior of building UI by the publisher

Reply:
Done